src/resource.rs

```diff
@@ -323,6 +417,42 @@ pub struct TypeDef {
     pub tag: TypeDefTag,
 }
 
+/// Concrete instantiations of types and operations defined in resources.
+pub trait CustomConcrete: sealed::SealedConcrete {
+    /// Name of the definition.
+    fn name(&self) -> &SmolStr;
+    /// Type arguments.
+    fn args(&self) -> &[TypeArg];
+    /// Parent resource.
+    fn resource(&self) -> &ResourceId;
```

Comment:
Note that there is a difference in meaning here - for ops, the resource means, you'll need access to that resource in order to do anything much to the op; as per resource checking. But for types, it doesn't really matter what the resource is - using a type from a resource doesn't mean you need the resource definition, or anything else, really; resource checking does not infer that an operation uses the resource just if it accepts/returns a type defined therein.

So again fine, but if you didn't have the trait, I don't think you need to pay as much attention to there being `CustomType::resource()` as `OpaqueOp::resource()`